pkg/parquetquery/iters.go

```diff
@@ -80,17 +80,41 @@ func EqualRowNumber(upToDefinitionLevel int, a, b RowNumber) bool {
 		(a[2] == b[2] || upToDefinitionLevel < 2) &&
 		(a[3] == b[3] || upToDefinitionLevel < 3) &&
 		(a[4] == b[4] || upToDefinitionLevel < 4) &&
-		(a[5] == b[5] || upToDefinitionLevel < 5)
+		(a[5] == b[5] || upToDefinitionLevel < 5) &&
+		(a[6] == b[6] || upToDefinitionLevel < 6) &&
+		(a[7] == b[7] || upToDefinitionLevel < 7)
 }
 
-func TruncateRowNumber(definitionLevelToKeep int, t RowNumber) RowNumber {
+func truncateRowNumberSlow(definitionLevelToKeep int, t RowNumber) RowNumber {
 	n := EmptyRowNumber()
 	for i := 0; i <= definitionLevelToKeep; i++ {
 		n[i] = t[i]
 	}
 	return n
 }
 
+func TruncateRowNumber(definitionLevelToKeep int, t RowNumber) RowNumber {
+	switch definitionLevelToKeep {
+	case 0:
+		return RowNumber{t[0], -1, -1, -1, -1, -1, -1, -1}
+	case 1:
+		return RowNumber{t[0], t[1], -1, -1, -1, -1, -1, -1}
+	case 2:
+		return RowNumber{t[0], t[1], t[2], -1, -1, -1, -1, -1}
+	case 3:
+		return RowNumber{t[0], t[1], t[2], t[3], -1, -1, -1, -1}
+	case 4:
+		return RowNumber{t[0], t[1], t[2], t[3], t[4], -1, -1, -1}
+	case 5:
+		return RowNumber{t[0], t[1], t[2], t[3], t[4], t[5], -1, -1}
+	case 6:
+		return RowNumber{t[0], t[1], t[2], t[3], t[4], t[5], t[6], -1}
+	case 7:
+		return RowNumber{t[0], t[1], t[2], t[3], t[4], t[5], t[6], t[7]}
+	}
+	return TruncateRowNumber(definitionLevelToKeep, t)
```

Comment:
Ah good catch! I meant to put the old truncateRowNumberSlow here but it might be good to log the error and return an empty row number instead.

Reply:
In my opinion logging and returning empty `RowNumber` is not a good way of error handling:
* It is not an idiomatic way to handle errors in Go
* Logging errors instead of handling them properly only delays discovery of bugs
* For the caller the function returns a potentially harmful value without returning an error. Callers might need to check for empty row number in order to detect "errors", which is not obvious for the caller

I think there are better ways to deal with invalid inputs here:
* Do the next best thing and treat `definitionLevelToKeep < 0` as 0 and `definitionLevelToKeep > MaxDefinitionLevel` as `MaxDefinitionLevel`. This is how the function behaved before.
* Simply panic (or log and panic). Code that calls `TruncateRowNumber()` is clearly a bug and should not happen at all. If someone adds definition levels to the schema in the future without increasing the size of `RowNumber` the bug is discovered quickly

Reply:
That's a great point. It did take me a very long time to figure out why the code wasn't working at 6 levels before. 👍